File: src/main/java/days/Day17.java
package days;

import setup.Day;
import util.Dijkstra;
import util.Direction;
import util.Grid;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.stream.Stream;

public class Day17 extends Day {

    //    private Graph<Grid.Tile<Num>> graph;
    private Grid<String> grid;

    @Override
    public void processInput() {
        grid = Grid.parseGrid(input);
//        graph = Grid.gridToGraph(grid);
    }

    @Override
    public Object part1() {
        var start = grid.getTile(0, 0);
        var end = grid.getTile(grid.width - 1, grid.height - 1);

        var d = Dijkstra.shortestPath(
                new State(start, Direction.E, 0, null),
                s -> s.position().equals(end),
                State::neighbors,
                (c, s) -> Integer.parseInt(s.position.data())
        );

        for (State state : d.end().path()) {
            System.out.println(state);
        }

        return d.get();
    }

    @Override
    public Object part2() {
        return null;
    }

    @Override
    public int getDay() {
        return 17;
    }

    @Override
    public boolean isTest() {
        return false;
    }

    @Override
    public String partOneSolution() {
        return "1238";
    }

    private record State(Grid.Tile<String> position, Direction dir, int stepsTakenInDir, State prevState) {
        public State move(Direction nDir) {
            return new State(position.add(nDir.asPoint()), nDir,   dir == nDir ? stepsTakenInDir + 1 : 1, this);
        }

        public boolean isValid() {
            return position != null && stepsTakenInDir <= 3;
        }

        public List<State> neighbors() {
            return Stream.of(
                    move(dir),
                    move(dir.left()),
                    move(dir.right())
            ).filter(State::isValid).toList();
        }

        public List<State> path() {
            var result = new ArrayList<State>();
            var current = this;
            while (current != null) {
                result.add(current);
                current = current.prevState;
            }
            Collections.reverse(result);
            return result;
        }

        @Override
        public boolean equals(Object o) {
            if (o == null || getClass() != o.getClass()) return false;
            State state = (State) o;
            return stepsTakenInDir == state.stepsTakenInDir && dir == state.dir && Objects.equals(position, state.position);
        }

        @Override
        public int hashCode() {
            return Objects.hash(position, dir, stepsTakenInDir);
        }

        @Override
        public String toString() {
            return "State{" +
                    "position=" + position +
                    ", dir=" + dir +
                    ", stepsTakenInDir=" + stepsTakenInDir +
                    '}';
        }
    }
}

File: src/days/day17.py
from collections import defaultdict

from src.setup.day import Day
from src.util.solution import solution

class Day17(Day):
    @property
    def day(self):
        return 17

    @solution("4372")
    def part1(self) -> object:
        containers = [int(x) for x in self.input]
        containers.sort(reverse=True)
        fit = 150

        def find_combinations(containers, fit):
            if fit == 0:
                return 1
            if fit < 0 or len(containers) == 0:
                return 0
            return find_combinations(containers[1:], fit) + find_combinations(containers[1:], fit - containers[0])

        return find_combinations(containers, fit)

    @solution("4")
    def part2(self) -> object:
        containers = [int(x) for x in self.input]
        containers.sort(reverse=True)
        fit = 150

        counts = defaultdict(int)

        def find_combinations(containers, fit, used_containers=[]):
            if fit == 0:
                counts[len(used_containers)] += 1
                return 1
            if fit < 0 or len(containers) == 0:
                return 0
            return find_combinations(containers[1:], fit, used_containers) + find_combinations(containers[1:], fit - containers[0], used_containers + [containers[0]])

        find_combinations(containers, fit)
        return counts[min(counts.keys())]

# Day17("test").run()
Day17().run()
